x-pack/plugin/ml-feature-index-builder/src/main/java/org/elasticsearch/xpack/ml/featureindexbuilder/job/FeatureIndexBuilderJobState.java

```diff
@@ -18,40 +19,67 @@
 import org.elasticsearch.xpack.core.indexing.IndexerState;
 
 import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.Objects;
+import java.util.TreeMap;
+
 import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.optionalConstructorArg;
 
 public class FeatureIndexBuilderJobState implements Task.Status, PersistentTaskState {
     public static final String NAME = "xpack/feature_index_builder/job";
 
     private final IndexerState state;
 
+    @Nullable
+    private final TreeMap<String, Object> currentPosition;
+
     private static final ParseField STATE = new ParseField("job_state");
+    private static final ParseField CURRENT_POSITION = new ParseField("current_position");
 
+    @SuppressWarnings("unchecked")
     public static final ConstructingObjectParser<FeatureIndexBuilderJobState, Void> PARSER = new ConstructingObjectParser<>(NAME,
-            args -> new FeatureIndexBuilderJobState((IndexerState) args[0]));
+            args -> new FeatureIndexBuilderJobState((IndexerState) args[0], (HashMap<String, Object>) args[1]));
 
     static {
         PARSER.declareField(constructorArg(), p -> {
             if (p.currentToken() == XContentParser.Token.VALUE_STRING) {
                 return IndexerState.fromString(p.text());
             }
             throw new IllegalArgumentException("Unsupported token [" + p.currentToken() + "]");
+
         }, STATE, ObjectParser.ValueType.STRING);
+        PARSER.declareField(optionalConstructorArg(), p -> {
+            if (p.currentToken() == XContentParser.Token.START_OBJECT) {
+                return p.map();
+            }
+            if (p.currentToken() == XContentParser.Token.VALUE_NULL) {
+                return null;
+            }
+            throw new IllegalArgumentException("Unsupported token [" + p.currentToken() + "]");
+        }, CURRENT_POSITION, ObjectParser.ValueType.VALUE_OBJECT_ARRAY);
     }
 
-    public FeatureIndexBuilderJobState(IndexerState state) {
+    public FeatureIndexBuilderJobState(IndexerState state, @Nullable Map<String, Object> position) {
         this.state = state;
+        this.currentPosition = position == null ? null : new TreeMap<>(position);
```

Comment:
If the intention is for this to be immutable then you should wrap the `new TreeMap` with `Collections.unmodifiableSortedMap()`, because otherwise a caller can modify it via the getter.